test/common/utxobased/db/NewProcessor.spec.ts

```diff
@@ -0,0 +1,322 @@
+/* eslint-disable @typescript-eslint/no-unused-expressions */
+import * as chai from 'chai'
+import { expect } from 'chai'
+import { makeMemoryDisklet } from 'disklet'
+
+import {
+  makeNewProcessor,
+  NewProcessor
+} from '../../../../src/common/utxobased/db/newProcessor'
+import {
+  IProcessorTransaction,
+  ITransactionInput,
+  ITransactionOutput
+} from '../../../../src/common/utxobased/db/types'
+
+chai.should()
+
+describe('Processor transactions tests', () => {
+  let processor: NewProcessor
+
+  function assertNumTransactions(expectedNum: number): void {
+    const num = processor.numTransactions()
+    expect(num).eql(expectedNum)
+  }
+
+  it('empty transaction baselets', async () => {
+    const storage = {}
+    const disklet = makeMemoryDisklet(storage)
+
+    processor = await makeNewProcessor({ disklet })
+    assertNumTransactions(0)
+  })
+
+  it('insert a transaction to transaction baselets', async () => {
+    const storage = {}
+    const disklet = makeMemoryDisklet(storage)
+    processor = await makeNewProcessor({ disklet })
+
+    const input1: ITransactionInput = {
+      txId: 'random',
+      outputIndex: 0,
+      scriptPubkey: 'pubkeyin1',
+      n: 0,
+      amount: '1'
+    }
+    const output1: ITransactionOutput = {
+      amount: '1',
+      n: 0,
+      scriptPubkey: 'pubkeyout1'
+    }
+    const output2: ITransactionOutput = {
+      amount: '1',
+      n: 1,
+      scriptPubkey: 'pubkeyout2'
+    }
+    const transaction1: IProcessorTransaction = {
+      txid: 'transaction1',
+      hex: '',
+      blockHeight: 1,
+      date: new Date(10).getTime(),
+      fees: '1',
+      inputs: [input1],
+      outputs: [output1, output2],
+      ourIns: [],
+      ourOuts: [],
+      ourAmount: '0'
+    }
+
+    await processor.insertTransaction({
+      tx: transaction1,
+      scriptPubkey: 'pubkeyout1'
+    })
+
+    assertNumTransactions(1)
+    const [tx1] = await processor.fetchTransactions({ txId: 'transaction1' })
+    expect(tx1.ourOuts[0]).to.eqls('0')
+    expect(tx1.ourAmount).to.eqls('1')
+    const [txByBlockHeight1] = await processor.fetchTransactions({
+      blockHeight: 1
+    })
+    expect(txByBlockHeight1.blockHeight).to.eqls(1)
+
+    // insert the same transaction, but with a script pubkey referencing an input
+    await processor.insertTransaction({
+      tx: transaction1,
+      scriptPubkey: 'pubkeyin1'
+    })
```

Comment:
We can insert to update?

Reply:
Yes, it does make somewhat sense from a network perspective. We have network calls, where the network does not know for sure if you already have the transaction, and we have some where we can be certain that the transaction is already saved.